File: src/controllers/notificationController.js
const Notification = require("../models/Notification");
const catchAsync = require("../utils/catchAsync");
const AppError = require("../utils/AppError");

const getNotifications = catchAsync(async (req, res) => {
  const {
    page,
    limit,
    sort,
    isRead
  } = req.query;

  const filter = {
    user: req.user._id
  };

  if (isRead !== undefined) {
    filter.isRead = isRead;
  }

  const skip = (page - 1) * limit;
  const [notifications, total] = await Promise.all([
    Notification.find(filter)
      .populate("relatedBorrowRecord")
      .sort(sort)
      .skip(skip)
      .limit(limit),
    Notification.countDocuments(filter)
  ]);

  res.status(200).json({
    status: "success",
    results: notifications.length,
    pagination: {
      page,
      limit,
      total,
      pages: Math.ceil(total / limit)
    },
    data: {
      notifications
    }
  });
});

const markAsRead = catchAsync(async (req, res, next) => {
  const notification = await Notification.findOneAndUpdate(
    {
      _id: req.params.id,
      user: req.user._id
    },
    {
      isRead: true,
      readAt: new Date()
    },
    {
      new: true,
      runValidators: true
    }
  );

  if (!notification) {
    return next(new AppError("Notification not found.", 404));
  }

  return res.status(200).json({
    status: "success",
    data: {
      notification
    }
  });
});

module.exports = {
  getNotifications,
  markAsRead
};

File: java-backend/NotificationController.java
import com.sun.net.httpserver.HttpExchange;

import java.io.IOException;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

class NotificationController {
  private final LibraryDataStore store;
  private final SecurityService security;
  private final JsonViews views;
  private final BorrowService borrowService;

  NotificationController(LibraryDataStore store, SecurityService security, JsonViews views, BorrowService borrowService) {
    this.store = store;
    this.security = security;
    this.views = views;
    this.borrowService = borrowService;
  }

  void notifications(HttpExchange exchange) throws IOException {
    User user = security.requireAuth(exchange);
    if (user == null) {
      return;
    }
    borrowService.markOverdueRecords();
    Map<String, String> query = RequestUtil.query(exchange);
    String isRead = query.getOrDefault("isRead", "");
    List<Notification> rows = store.notifications().stream()
      .filter(notification -> "librarian".equals(user.role) || notification.userId.equals(user.id))
      .collect(Collectors.toList());
    if (!isRead.isBlank()) {
      boolean read = Boolean.parseBoolean(isRead);
      rows = rows.stream().filter(notification -> notification.read == read).collect(Collectors.toList());
    }
    rows.sort(Comparator.comparing((Notification notification) -> notification.createdAt).reversed());
    int page = Math.max(1, TextUtil.parseInt(query.getOrDefault("page", "1"), 1));
    int limit = Math.min(100, Math.max(1, TextUtil.parseInt(query.getOrDefault("limit", "20"), 20)));
    sendNotifications(exchange, rows, page, limit);
  }

  void readNotification(HttpExchange exchange, String notificationId) throws IOException {
    User user = security.requireAuth(exchange);
    if (user == null) {
      return;
    }
    Notification notification = store.findNotification(notificationId);
    if (notification == null || ("member".equals(user.role) && !notification.userId.equals(user.id))) {
      HttpResponses.sendError(exchange, 404, "Notification not found.");
      return;
    }
    notification.read = true;
    notification.updatedAt = TimeUtil.now();
    HttpResponses.sendJson(exchange, 200, "{\"status\":\"success\",\"data\":{\"notification\":" + views.notification(notification) + "}}");
  }

  private void sendNotifications(HttpExchange exchange, List<Notification> rows, int page, int limit) throws IOException {
    int total = rows.size();
    int pages = Math.max(1, (int) Math.ceil(total / (double) limit));
    int start = Math.min(total, (page - 1) * limit);
    int end = Math.min(total, start + limit);
    List<Notification> pageRows = rows.subList(start, end);
    String pagination = JsonUtil.pagination(page, limit, total, pages);
    HttpResponses.sendJson(exchange, 200, "{\"status\":\"success\",\"results\":" + pageRows.size() + ",\"pagination\":" + pagination + ",\"data\":{\"notifications\":" + JsonUtil.array(pageRows, views::notification) + "}}");
  }
}
